src/main/java/io/openliberty/tools/common/plugins/util/DevUtil.java

```diff
@@ -2422,22 +2423,38 @@ private void toggleFeatureGeneration() {
         info("Setting automatic generation of features to: " + getFormattedBooleanString(generateFeatures));
         if (generateFeatures) {
             // If hotkey is toggled to “true”, generate features right away.
-            optimizeFeatures();
+            optimizeGenerateFeatures();
         }
     }
 
     /**
      * Generate features using all classes and only user specified features.
      */
-    private void optimizeFeatures() {
+    private void optimizeGenerateFeatures() {
         info("Generating optimized features list...");
+        // scan all class files and provide only user specified features
         try {
-            libertyGenerateFeatures();
+            libertyGenerateFeatures(null, true);
         } catch (PluginExecutionException e) {
             error("An error occurred while trying to optimize features: " + e.getMessage(), e);
         }
     }
 
+     /**
+     * Generate features using updated classes and all existing features.
+     */
+    private void incrementGenerateFeatures() {
+        info("Generating feature list from incremental changes...");
+        try {
+            Collection <String> javaSourceClassPaths = getClassPaths(javaSourceClasses);
+            libertyGenerateFeatures(javaSourceClassPaths, false);
+        } catch (PluginExecutionException | IOException e) {
+            error("An error occurred while trying to generate features: " + e.getMessage(), e);
+        }
+        // TODO: update clear logic to handle cases where feature generation fails
+        javaSourceClasses.clear();
```

Comment:
It looks like if feature generation fails `javaSourceClasses` will still be cleared. Wondering if this line should be after 2450:
```java
            libertyGenerateFeatures(javaSourceClassPaths, false);
            javaSourceClasses.clear();
        } catch (PluginExecutionException | IOException e) {
```

Reply:
Maybe on error feature generation should be turned off?

Reply:
Both of these suggestions are good. I plan to address dealing with the error/exception scenarios in another PR. I'm still not clear on the flow for the feature generation goal failing (ie. when we fail it on our terms vs something else goes wrong). I was thinking of catching that earlier on and maybe having a return value for `libertyGenerateFeatures()`. So if an Exception occurs here, it will be one that was not expected to be caught. So the second suggestion may work well here.